runtimevar/etcdvar/etcdvar.go

```diff
@@ -70,84 +62,106 @@ func (s *state) UpdateTime() time.Time {
 	return s.updateTime
 }
 
-// errorState returns a new State with err, unless prevS also represents
-// the same error, in which case it returns nil.
-func errorState(err error, prevS driver.State) driver.State {
-	s := &state{err: err}
-	if prevS == nil {
-		return s
-	}
-	prev := prevS.(*state)
-	if prev.err == nil {
-		// New error.
-		return s
-	}
-	if err == prev.err {
-		return nil
-	}
-	var code, prevCode codes.Code
-	if etcdErr, ok := err.(rpctypes.EtcdError); ok {
-		code = etcdErr.Code()
-	}
-	if etcdErr, ok := prev.err.(rpctypes.EtcdError); ok {
-		prevCode = etcdErr.Code()
-	}
-	if code != codes.OK && code == prevCode {
-		return nil
-	}
-	return s
-}
-
 // watcher implements driver.Watcher.
 type watcher struct {
-	name    string
-	client  *clientv3.Client
-	decoder *runtimevar.Decoder
-	wait    time.Duration
+	// The background goroutine writes new *state values to ch.
+	// It is buffered so that the background goroutine can write without
+	// blocking; it always drains the buffer before writing so that the latest
+	// write is buffered. If writes could block, the background goroutine could be
+	// blocked indefinitely from reading etcd's Watch events.
+	ch chan *state
```

Comment:
Can these two channels be combined? When the watch is canceled, it closes ch?

Reply:
Yup, good catch. That works; done.